lightning/src/ln/channel.rs

```diff
@@ -1435,6 +1437,223 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+
+	fn dual_funding_context(&self) -> &DualFundingChannelContext;
+
+	fn dual_funding_context_mut(&mut self) -> &mut DualFundingChannelContext;
+
+	fn is_initiator(&self) -> bool;
+
+	fn begin_interactive_funding_tx_construction<ES: Deref>(
+		&mut self, entropy_source: &ES, holder_node_id: PublicKey,
+	) -> Result<Option<InteractiveTxMessageSend>, APIError>
+	where ES::Target: EntropySource
+	{
+		let mut funding_outputs = Vec::new();
+		if self.is_initiator() {
+			let tx_out = TxOut {
+				value: Amount::from_sat(self.context().get_value_satoshis()),
+				script_pubkey: self.context().get_funding_redeemscript().to_p2wsh(),
+			};
+			funding_outputs.push(
+				if self.dual_funding_context_mut().their_funding_satoshis.unwrap_or(0) > 0 {
+					OutputOwned::Shared(SharedOwnedOutput::new(tx_out, self.dual_funding_context_mut().our_funding_satoshis))
+				} else {
+					OutputOwned::SharedControlFullyOwned(tx_out)
+				}
+			);
+		}
+
+		// if let Ok((tx_constructor, msg)) = 
+		match InteractiveTxConstructor::new(
+			entropy_source, self.context().channel_id(),
+			self.dual_funding_context_mut().funding_feerate_sat_per_1000_weight,
+			holder_node_id, self.context().counterparty_node_id, self.is_initiator(),
+			self.dual_funding_context_mut().funding_tx_locktime,
+			self.dual_funding_context_mut().our_funding_inputs.clone().unwrap_or(vec![]), funding_outputs,
+			Some((self.context().get_funding_redeemscript(), self.context().channel_value_satoshis))
+		) {
+			Ok((tx_constructor, msg)) => {
+				self.interactive_tx_constructor_mut().replace(tx_constructor);
+				Ok(msg)
+			},
+			Err(err) => {
+				Err(APIError::APIMisuseError { err:
+					match err {
+				        AbortReason::MissingFundingOutput => "Funding output missing".into(),
+				        AbortReason::DuplicateFundingOutput => "Duplicate funding output".into(),
+				        AbortReason::InvalidLowFundingOutputValue => "Local part of funding output is greater than the output value".into(),
+				        _ => {
+				        	"Error starting interactive transaction construction".into()
+				        }
+			    	}
+			    })
+			}
+		}
+	}
+
+	fn tx_add_input(&mut self, msg: &msgs::TxAddInput) -> InteractiveTxMessageSendResult {
+		InteractiveTxMessageSendResult(match self.interactive_tx_constructor_mut() {
+			Some(ref mut tx_constructor) => tx_constructor.handle_tx_add_input(msg).map_err(
+				|reason| reason.into_tx_abort_msg(self.context_mut().channel_id())),
+			None => Err(msgs::TxAbort {
+				channel_id: self.context_mut().channel_id(),
+				data: "We do not have an interactive transaction negotiation in progress".to_string().into_bytes()
+			}),
+		})
+	}
+
+	fn tx_add_output(&mut self, msg: &msgs::TxAddOutput)-> InteractiveTxMessageSendResult {
+		InteractiveTxMessageSendResult(match self.interactive_tx_constructor_mut() {
+			Some(ref mut tx_constructor) => tx_constructor.handle_tx_add_output(msg).map_err(
+				|reason| reason.into_tx_abort_msg(self.context_mut().channel_id())),
```

Comment:
Context can be non-mutable (`context()`) in a few other places as well

Reply:
Yip. Think this was a search and replace issue at some stage in a prior version.